pkg/updatestrategy/aws_ec2.go

```diff
@@ -269,3 +273,93 @@ func (n *EC2NodePoolBackend) decommission(ctx context.Context, filters []*ec2.Fi
 		}
 	}
 }
+
+type KarpenterCRDNameResolver struct {
+	NodePoolCRDName string
+	k8sClients      *kubernetes.ClientsCollection
+}
+
+func NewKarpenterCRDResolver(ctx context.Context, k8sClients *kubernetes.ClientsCollection) (*KarpenterCRDNameResolver, error) {
+	exists, err := k8sClients.Exists(ctx, crd, "", karpenterNodePoolResource, v1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		return &KarpenterCRDNameResolver{
+			NodePoolCRDName: karpenterNodePoolResource,
+			k8sClients:      k8sClients,
+		}, nil
+	}
+	return &KarpenterCRDNameResolver{
+		NodePoolCRDName: karpenterProvisionerResource,
+		k8sClients:      k8sClients,
+	}, nil
+}
+
+func (r *KarpenterCRDNameResolver) NodeTemplateCRDName() string {
+	switch r.NodePoolCRDName {
+	case karpenterNodePoolResource:
+		return KarpenterEC2NodeClassResource
+	default:
+		return karpenterAWSNodeTemplateResource
+	}
+}
+
+func (r *KarpenterCRDNameResolver) referenceFieldName() string {
+	switch r.NodePoolCRDName {
+	case karpenterNodePoolResource:
+		return "nodeClassRef"
+	default:
+		return "providerRef"
+	}
+}
+
+func (r *KarpenterCRDNameResolver) getAMIsFromSpec(spec interface{}) string {
+	switch r.NodePoolCRDName {
+	case karpenterNodePoolResource:
+		amiSelectorTerms := spec.(map[string]interface{})["amiSelectorTerms"].([]interface{})
+		var amis []string
+		for _, amiSelectorTerm := range amiSelectorTerms {
+			if amiSelectorTerm.(map[string]interface{})["id"] != nil {
+				amis = append(amis, amiSelectorTerm.(map[string]interface{})["id"].(string))
+			}
+		}
+		return strings.Join(amis, ",")
+	default:
+		return spec.(map[string]interface{})["amiSelector"].(map[string]interface{})["aws-ids"].(string)
+	}
+}
+
+func (r *KarpenterCRDNameResolver) getInstanceTag() string {
+	switch r.NodePoolCRDName {
+	case karpenterNodePoolResource:
+		return karpenterNodePoolTag
+	default:
+		return karpenterProvisionerTag
+	}
+}
+
+func (r *KarpenterCRDNameResolver) NodePoolConfigGetter(ctx context.Context, nodePool *api.NodePool) (*InstanceConfig, error) {
+	// CLM assumes that the node pool name is used for both the node-pool and the node-template that it references
+	NodeTemplate, err := r.k8sClients.Get(ctx, r.NodeTemplateCRDName(), "", nodePool.Name, v1.GetOptions{})
+	//if apierrors.IsNotFound(err) {
+	//	// the node pool have been deleted. thus the
+	//	return nil, nil
+	//}
```

Comment:
nit: remove commented code

Reply:
🤔 
actually, this may manage to mark old karpenter node-pools for decommission.
I will experiment with uncomment it and keep them to see if it makes a difference